logstreams/src/main/java/io/zeebe/distributedlog/impl/DefaultDistributedLogstreamService.java

```diff
@@ -263,72 +245,58 @@ public void backup(BackupOutput backupOutput) {
   @Override
   public void restore(BackupInput backupInput) {
     final long backupPosition = backupInput.readLong();
-    restore(backupPosition);
-
-    logger.debug("Restored local log to position {}", lastPosition);
-    currentLeader = backupInput.readString();
-    currentLeaderTerm = backupInput.readLong();
-  }
 
-  public void restore(long backupPosition) {
     if (lastPosition < backupPosition) {
       LogstreamConfig.startRestore(localMemberId, partitionId);
+      final RestoreFactory restoreFactory = LogstreamConfig.getRestoreFactory(localMemberId);
+
+      final ThreadContext restoreThreadContext =
+          new SingleThreadContext(String.format("log-restore-%d", partitionId));
+      final RestoreClient restoreClient = restoreFactory.createClient(partitionId);
+      final RestoreNodeProvider nodeProvider = restoreFactory.createNodeProvider(partitionId);
+      final LogReplicator logReplicator =
+          new LogReplicator(this, restoreClient, restoreThreadContext);
+      final SnapshotRequester snapshotReplicator =
+          new SnapshotRequester(
+              restoreClient, restoreFactory.createSnapshotRestoreContext(), partitionId);
+      final RestoreController restoreController =
+          new RestoreController(
+              restoreClient,
+              nodeProvider,
+              logReplicator,
+              snapshotReplicator,
+              restoreThreadContext,
+              getLogger());
 
-      // pick the correct restore strategy and execute forever until the log is restored, e.g.
-      // lastPosition >= backupPosition.
       while (lastPosition < backupPosition) {
         final long latestLocalPosition = lastPosition;
-        logger.debug("Restoring local log from position {} to {}", lastPosition, backupPosition);
+        logger.trace(
+            "Restoring local log from position {} to {}", latestLocalPosition, backupPosition);
+
         try {
-          final DefaultStrategyPicker strategyPicker = this.buildRestoreStrategyPicker();
-          final long lastUpdatedPosition =
-              strategyPicker
-                  .pick(latestLocalPosition, backupPosition)
-                  .thenCompose(RestoreStrategy::executeRestoreStrategy)
-                  .join();
-          updateCommitPosition(lastUpdatedPosition);
+          lastPosition = restoreController.restore(lastPosition, backupPosition);
```

Comment:
This is safe, right, since we always replicate at least one event so the append method always updates the `logStream` commit position?

Reply:
Yes.